azerothcore/azerothcore-wotlk
# refactor(Core/sLog): improve singleton management

## CHANGES PROPOSED:
- Created [interface](https://www.tutorialspoint.com/cplusplus/cpp_interfaces.htm) for the `Log` class in a separate `ILog.h` file
- Move all `enum` blocks from `Log.h` to `ILog.h`
- Refactored `sLog` singleton to use `std::unique_ptr` (thanks @Rochet2)
- Fixed CMake issue in tests and add the `mock` directory (thanks @Yehonal)
- Fixed linking issue in tests (thanks @jackpoz)
- Created `LogMock` class to be used in unit tests instead of `Log`
- Added tests for `BaseGain` of `Formulas.h`

## ISSUES ADDRESSED:
- No reported issue
- The same problem described in [this article](https://helpercode.com/2015/09/16/how-to-fake-a-singleton-in-c/), but with a different solution



## TESTS PERFORMED:
- Compile, run unit tests with and without mocking the Log: the output is as expected.

- Also the CI already tested the `sLog` singleton still works as intended:
 
![image](https://user-images.githubusercontent.com/75517/101218322-2851d780-3683-11eb-8356-9e22b2ee4263.png)

## How to test the changes

- A quick check in game is more than enough, also have a look at your `worldserver` window and make sure it still outputs messages correctly.

## Target branch(es):








 
 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
